CASSANDRA-18511 Add support for JMX in the in-jvm dtest framework (3.11)

JeetKunDoug opened the pull request.

JeetKunDoug committed:
CASSANDRA-18511 Add support for JMX in the in-jvm dtest framework

JeetKunDoug committed:
Use snapshot of in-jvm dtest API and bump resources for Circle

jonmeredith commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -33,62 +38,144 @@
+            return new NoOpMBeanWrapper();
+
+        String klass = System.getProperty(MBEAN_REGISTRATION_CLASS);
+        if (klass == null)
+            if (Boolean.getBoolean(IS_IN_JVM_DTEST))
+                return new NoOpMBeanWrapper();
+            else
+                return new PlatformMBeanWrapper();
```
nit: my personal preference would be to include the outer `if` braces `{}` but don't think this goes against the style guide.

JeetKunDoug replied:
Yeah - I get dinged for _including_ them so I try not to do it if it's only single-lined.

JeetKunDoug replied:
Rereading the style guide, it does specifically say they _may_ be elided so I'm actually going to put them in, as I generally agree with you on this.

Actually, the entire rest of the file has elided the braces for single-lines, so I'm going to leave it alone.

jonmeredith commented on test/distributed/org/apache/cassandra/distributed/test/jmx/JMXFeatureTest.java:
```diff
@@ -0,0 +1,94 @@
+        // NOTE: At some point, the hostname of the broadcastAddress can be resolved
+        // and then the `getHostString`, which would otherwise return the IP address,
+        // starts returning `localhost` - use `.getAddress().getHostAddress()` to work around this.
+        String jmxHost = instance.config().broadcastAddress().getAddress().getHostAddress();
+        int jmxPort = instance.config().jmxPort();
+        String url = String.format(JMX_SERVICE_URL_FMT, jmxHost, jmxPort);
+        try (JMXConnector jmxc = JMXConnectorFactory.connect(new JMXServiceURL(url), null))
```
is it worth making a convenience method that takes an IInstanceConfig and returns the `JMXConnector`. Is there any test where you wouldn't do those same four statements to make the connector?

JeetKunDoug replied:
So would this be worth putting into the dtest API itself in that case? Unfortunately would require another round of review/release of the dtest API but everything you need to get a connection is available in the API.

Cleanup is probably easier if we just do it in the Cassandra code, as you could, in theory, stash the JMXConnector on the instance itself, and then close it when the instance is closed, which would be easier to enforce.

JeetKunDoug replied:
We pulled this into the dtest framework.

jonmeredith commented on test/distributed/org/apache/cassandra/distributed/test/jmx/JMXGetterCheckTest.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * This class is meant to make new errors easier to read, by adding the JMX endpoint, and cleaning up the unneded JMX/Reflection logic cluttering the stacktrace
```
```suggestion
     * This class is meant to make new errors easier to read, by adding the JMX endpoint, and cleaning up the unneeded JMX/Reflection logic cluttering the stacktrace
```

JeetKunDoug replied:
Was actually existing code (in the other branches), but I'll fix the typo anyway while I'm at it.

JeetKunDoug committed:
Fix additional leak in jmx by decreasing long lease/keep-alive defaults in JMX.

Additionally, added the ability to run some code in the looper tests
that's passed from the test itself so that it can better exercise functionality.

# Conflicts:
#	relocate-dependencies.pom
#	src/java/org/apache/cassandra/config/CassandraRelevantProperties.java
#	src/java/org/apache/cassandra/utils/MBeanWrapper.java
#	test/distributed/org/apache/cassandra/distributed/impl/Instance.java
#	test/distributed/org/apache/cassandra/distributed/test/ResourceLeakTest.java

JeetKunDoug committed:
Use JMXUtil from new dtest-api in tests.

JeetKunDoug committed:
Typo and import order fixes

JeetKunDoug committed:
Moved jmx handling out of `Instance` into its own class

Also fixed a nit in the property name for in-jvm dtests

JeetKunDoug committed:
Fix formatting in MBeanWrapper

JeetKunDoug committed:
Add TEST prefix to in-jvm dtest prop

JeetKunDoug committed:
Rename prop after futher discussion

JeetKunDoug committed:
Fix missing check in GCInspector

JeetKunDoug committed:
Fix shutdown bug in refactored JMX

JeetKunDoug committed:
Address review comments on mBean vs. MBean and rename getMbs to getMBeanServer

JeetKunDoug committed:
Use released dtest-api artifact
